src/simd.rs

```diff
@@ -0,0 +1,198 @@
+//! SIMD byte scanning logic.
+//! Inspired by this great overview: http://0x80.pl/articles/simd-byte-lookup.html
+//! However, since all of the bytes we're interested in, we don't quite need the
+//! fully generality of the universal algorithm and are hence able to skip a few
+//! instructions.
+
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+use core::arch::x86_64::*;
+
+pub(crate) enum LoopInstruction<T> {
+    /// Continue looking for more special bytes, but skip next few bytes.
+    ContinueAndSkip(usize),
+    /// Break looping immediately, returning with the given index and value.
+    BreakAtWith(usize, T)
+}
+
+#[allow(overflowing_literals)]
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+unsafe fn compute_mask(bytes: &[u8], ix: isize) -> i32 {
+    // constants. computed with the below code
+    // ```rust
+    // let chars = [b'\n', b'\r', b'*', b'_', b'~', b'|', b'&', b'\\', b'[', b']', b'<', b'!', b'`'];
+    // let mut lower_bitmap = [0u8; 16];
+    // for &c in &chars {
+    //     lower_bitmap[(c & 0x0f) as usize] |= 1 << (c >> 4);
+    // }
+    // let bitmap = unsafe { _mm_loadu_si128(lower_bitmap.as_ptr() as *const _) };
+    // ```
+    let bitmap = _mm_setr_epi8(
+        64, 4, 0, 0, 0, 0, 4, 0, 0, 0, 5, 32, 168, 33, 128, 32
+    );
+    let bitmask_lookup = _mm_setr_epi8(
+        1, 2, 4, 8, 16, 32, 64, -128,
+        1, 2, 4, 8, 16, 32, 64, -128
+    );
+
+    // actual computation
+    let raw_ptr = bytes.as_ptr().offset(ix) as *const _;
+    let input = _mm_loadu_si128(raw_ptr);
+    let bitset = _mm_shuffle_epi8(bitmap, input);
```

Comment:
One of the problems we have here is that this is only available when the CPU has "ssse3" feature ([docs](https://doc.rust-lang.org/beta/core/arch/x86_64/fn._mm_shuffle_epi8.html)), and this is not guaranteed. For this code to be safe, there needs to be a runtime check. I haven't audited all such uses.

Reply:
Woops - hadn't realized that. Will fix, thanks!